### PR TITLE
Fix vector dimension type

### PR DESCRIPTION
**Pull request by QuerthDP**
Switch vector dimension handling to fixed-width `uint32_t` type, update parsing/validation, and add boundary tests.

The dimension is parsed as `unsigned long` at first which is guaranteed
to be **at least** 32-bit long, which is safe to downcast to `uint32_t`.

Fixes: SCYLLADB-223

### CODE REVIEW

File: cql3/Cql.g [hunk 1]
```diff
@@ -2071,7 +2072,12 @@ vector_type returns [shared_ptr<cql3::cql3_type::raw> pt]
         {
             if ($d.text[0] == '-')
                 throw exceptions::invalid_request_exception("Vectors must have a dimension greater than 0");
-            $pt = cql3::cql3_type::raw::vector(t, std::stoul($d.text));
+            auto parsed_dimension = std::stoul($d.text);
```
Comment (nyh) [inline cql3/Cql.g]: As far as I know, std::stoul() throws exceptions in some cases, so I think you should catch those and return a clean invalid_return_exception exception instead of some ugly "internal server error" due to an uncaught exception.

Please add a test for this case, checking cases like negative number (already caught by the previous), zero (is zero allowed? It seems it is! Is that deliberate), and also non-numeric garbage like "dog" or "123x" - and checking we get the expected error in such error cases.

File: cql3/Cql.g [hunk 2]
```diff
@@ -2071,7 +2072,12 @@ vector_type returns [shared_ptr<cql3::cql3_type::raw> pt]
...
+            static_assert(sizeof(unsigned long) >= sizeof(vector_dimension_t));
+            if (parsed_dimension > std::numeric_limits<vector_dimension_t>::max()) {
+                throw exceptions::invalid_request_exception("Vector dimension exceeds the type bounds");
```
Comment (nyh) [inline cql3/Cql.g]: By the way, I see elsewhere we have MAX_VECTOR_DIMENSION which gives a much lower maximum. Why don't we test this here instead of the ridiculously large max(), and provide a clearer error message about what the maximum size may be?

File: db/marshal/type_parser.cc [hunk 3]
```diff
@@ -142,7 +143,12 @@ std::tuple<data_type, size_t> type_parser::get_vector_parameters()
             throw parse_exception(_str, _idx, "expected digit or ')'");
         }
 
-        size = std::stoul(_str.substr(i, _idx - i));
+        auto parsed_size = std::stoul(_str.substr(i, _idx - i));
+        static_assert(sizeof(unsigned long) >= sizeof(vector_dimension_t));
+        if (parsed_size > std::numeric_limits<vector_dimension_t>::max()) {
```
Comment (nyh) [inline db/marshal/type_parser.cc]: Again, I wonder if we have MAX_VECTOR_DIMENSION why we don't check it here too.

File: db/marshal/type_parser.cc [hunk 4]
```diff
@@ -142,7 +143,12 @@ std::tuple<data_type, size_t> type_parser::get_vector_parameters()
             throw parse_exception(_str, _idx, "expected digit or ')'");
         }
 
-        size = std::stoul(_str.substr(i, _idx - i));
+        auto parsed_size = std::stoul(_str.substr(i, _idx - i));
```
Comment (nyh) [inline db/marshal/type_parser.cc]: Again, stoul() can throw, I wonder if we shouldn't catch it and rethrow a parse_exception() like we do in other failure cases.

File: test/cqlpy/test_type_vector.py [hunk 5]
```diff
@@ -26,3 +28,14 @@ def test_vector_of_set_using_arguments_binding(cql, test_keyspace):
...
         assert row.v == value_to_insert
+
+
+def test_vector_dimension_upper_bound_is_allowed(cql, test_keyspace, scylla_only):
+    with new_test_table(cql, test_keyspace, "pk int primary key, v vector<float, 16000>"):
+        pass
+
+
+def test_vector_dimension_above_upper_bound_is_rejected(cql, test_keyspace, scylla_only):
+    with pytest.raises(Exception, match="dimension"):
```
Comment (nyh) [inline test/cqlpy/test_type_vector.py]: I think it will be better to check for a more explicit exception type, probably InvalidRequest. We don't it to be a "no hosts available" or similar error with just the word "dimension" somewhere in the message.

File: test/cqlpy/test_type_vector.py [hunk 6]
```diff
@@ -26,3 +28,14 @@ def test_vector_of_set_using_arguments_binding(cql, test_keyspace):
 
         assert row is not None
         assert row.v == value_to_insert
+
+
+def test_vector_dimension_upper_bound_is_allowed(cql, test_keyspace, scylla_only):
```
Comment (nyh) [inline test/cqlpy/test_type_vector.py]: I would have added a comment here saying that this "16,000" is an artificial limit (MAX_VECTOR_DIMENSION) currently imposed by Scylla chosen because of (??), and Cassandra has a different limit. Otherwise, in the future readers might not know why this test focuses on the number 16000.

You can also make these test pass on Cassandra if you add something like
```
MAX_VECTOR_DIMENSION = 16000 if is_scylla(cql) else 1234 # whatever is the Cassandra limit
````

File: cql3/Cql.g (hunk 2 above)
Comment (QuerthDP) [reply cql3/Cql.g]: Should we duplicate the check, which is done in `prepare_internal`?

```
if (_dimension == 0) {
    throw exceptions::invalid_request_exception("Vectors must have a dimension greater than 0");
}

if (_dimension > MAX_VECTOR_DIMENSION) {
    throw exceptions::invalid_request_exception(
            format("Vectors must have a dimension less than or equal to {}", MAX_VECTOR_DIMENSION));
}
```

File: test/cqlpy/test_type_vector.py (hunk 6 above)
Comment (QuerthDP) [reply test/cqlpy/test_type_vector.py]: Added and descirbed

File: cql3/Cql.g (hunk 2 above)
Comment (QuerthDP) [reply cql3/Cql.g]: Added the check here as well

File: db/marshal/type_parser.cc (hunk 3 above)
Comment (QuerthDP) [reply db/marshal/type_parser.cc]: I thought the idea in this patch was to check the type boundaries, because the MAX_VECTOR_DIMENSION was already checked there. But surely we may check that earlier, which would allow us to terminate earlier with invalid request already

File: cql3/Cql.g (hunk 1 above)
Comment (QuerthDP) [reply cql3/Cql.g]: Good point, added a check on the parser level.

Added the tests

File: cql3/Cql.g (hunk 2 above)
Comment (nyh) [reply cql3/Cql.g]: I don't know. I guess it's not a good idea to have the same check in two places, but I don't know what is the "proper" place.

File: cql3/Cql.g [hunk 7]
```diff
@@ -2071,7 +2071,18 @@ vector_type returns [shared_ptr<cql3::cql3_type::raw> pt]
...
+            try {
+                parsed_dimension = std::stoul($d.text);
+            } catch (const std::exception& e) {
+                throw exceptions::invalid_request_exception(format("Invalid vector dimension: {}", $d.text));
+            }
+            static_assert(sizeof(unsigned long) >= sizeof(vector_dimension_t));
+            if (parsed_dimension > cql3::cql3_type::MAX_VECTOR_DIMENSION) {
+                throw exceptions::invalid_request_exception(
+                        format("Vectors must have a dimension less than or equal to {}", cql3::cql3_type::MAX_VECTOR_DIMENSION));
+            }
```
Comment (nyh) [inline cql3/Cql.g:2087]: I don't see here a check for ==0 being forbidden, but I guess this remains in the original validation function? I'm not sure which of these parsing function is called in which case.

File: cql3/cql3_type.hh [hunk 8]
```diff
@@ -39,6 +39,9 @@ public:
     data_type get_type() const { return _type; }
     const sstring& to_string() const { return _type->cql3_type_name(); }
 
+    // This limitation is acquired from the maximum number of dimensions in OpenSearch.
+    static constexpr vector_dimension_t MAX_VECTOR_DIMENSION = 16000;
```
Comment (nyh) [inline cql3/cql3_type.hh:43]: nitpick: Maybe "inline constexpr" is better than "static constexpr"?

File: db/marshal/type_parser.cc [hunk 9]
```diff
@@ -142,7 +143,17 @@ std::tuple<data_type, size_t> type_parser::get_vector_parameters()
...
-        size = std::stoul(_str.substr(i, _idx - i));
+        unsigned long parsed_size;
+        try {
+            parsed_size = std::stoul(_str.substr(i, _idx - i));
+        } catch (const std::exception& e) {
+            throw parse_exception(_str, i, format("Invalid vector dimension: {}", e.what()));
+        }
+        static_assert(sizeof(unsigned long) >= sizeof(vector_dimension_t));
+        if (parsed_size > cql3::cql3_type::MAX_VECTOR_DIMENSION) {
+            throw parse_exception(_str, _idx, format("Vectors must have a dimension less than or equal to {}", cql3::cql3_type::MAX_VECTOR_DIMENSION));
```
Comment (nyh) [inline db/marshal/type_parser.cc:157]: Again, missing here a check that 0 is not allowed, I hope it's somewhere else.

File: cql3/cql3_type.hh (hunk 8 above)
Comment (QuerthDP) [reply cql3/cql3_type.hh:43]: According to Gemini:

>For class members like this one, `static constexpr` is the correct choice. `inline constexpr` is for namespace-scope constants, not class members.
>
>`static constexpr` (class member): Makes it a class-level constant shared across all instances. Since C++17, these are implicitly inline, so no out-of-class definition is needed.

Furthermore, the `inline constexpr` won't even compile:
```
Non-static data member cannot be constexpr; did you intend to make it static? (fix available)clang(invalid_constexpr_member)
```

File: cql3/Cql.g (hunk 7 above)
Comment (QuerthDP) [reply cql3/Cql.g:2087]: It is checked in `cql3_type::raw_vector::prepare_internal`
```
virtual cql3_type prepare_internal(const sstring& keyspace, const data_dictionary::user_types_metadata& user_types) override {
    if (!is_frozen()) {
        freeze();
    }

    if (_dimension == 0) {
        throw exceptions::invalid_request_exception("Vectors must have a dimension greater than 0");
    }

    if (_dimension > MAX_VECTOR_DIMENSION) {
        throw exceptions::invalid_request_exception(
                format("Vectors must have a dimension less than or equal to {}", MAX_VECTOR_DIMENSION));
    }

    return vector_type_impl::get_instance(_type->prepare_internal(keyspace, user_types).get_type(), _dimension)->as_cql3_type();
}
```
It must be called anyway when creating a vector type.

File: db/marshal/type_parser.cc (hunk 9 above)
Comment (QuerthDP) [reply db/marshal/type_parser.cc:157]: As for here, according to CoPilot analysis, there might be possible that we don't execute `cql3_type::raw_vector::prepare_internal` afterwards. Those are different code paths, this one lies nearby sstable deserialization. The data here should be validated, but let's keep that strict. I'll add an additional check here as it's necessary

File: cql3/Cql.g (hunk 7 above)
Comment (QuerthDP) [reply cql3/Cql.g:2087]: Although I think, adding it here to be more explicit is also worth it. Why would be error out early only on incomplete set of edge cases?

There's a question then whether should we leave the checks in `prepare_internal` - yet I think we do, as it's the code which is executed f.e. in test not using CQL.

File: cql3/cql3_type.hh (hunk 8 above)
Comment (nyh) [reply cql3/cql3_type.hh:43]: I missed that it's a class-member. I misread static as file-local. Sorry about the spam.

**Review by nyh — APPROVED** [review]
